src/dotty/tools/dotc/transform/ElimErasedValueType.scala

```diff
@@ -0,0 +1,97 @@
+
+  override def transform(ref: SingleDenotation)(implicit ctx: Context): SingleDenotation = {
+    val info = ref match {
+      case ref: ClassDenotation if ref is ModuleClass =>
+        ref.linkedClass match {
+          case origClass: ClassSymbol if isDerivedValueClass(origClass) =>
+            val cinfo = ref.classInfo
+            val decls1 = cinfo.decls.cloneScope
+            ctx.atPhase(this.next) { implicit ctx =>
+              // Remove synthetic cast methods introduced by ExtensionMethods,
+              // they are no longer needed after this phase.
+              decls1.unlink(cinfo.decl(nme.UNDERLYING2EVT).symbol)
+              decls1.unlink(cinfo.decl(nme.EVT2UNDERLYING).symbol)
+            }
+            cinfo.derivedClassInfo(decls = decls1)
+          case _ =>
+            ref.info
+        }
+      case _ =>
+        ref.info
+    }
+    val info1 = elimErasedValueType(info)
+    if (info1 eq ref.info) ref
+    else ref match {
+      case ref: SymDenotation => ref.copySymDenotation(info = info1)
+      case _ => ref.derivedSingleDenotation(ref.symbol, info1)
+    }
+  }
+
+  override def checkPostCondition(tree: tpd.Tree)(implicit ctx: Context) = {
+    val tp = tree.tpe
+    val erased = elimErasedValueType(tp)
+    assert(tp eq erased,
+      i"The type $tp of tree $tree contains ErasedValueType, it should have been erased to $erased by ElimErasedValueType.")
+    val name = tree.symbol.name
+    assert(name != nme.UNDERLYING2EVT && name != nme.EVT2UNDERLYING,
+      i"The symbol '${tree.symbol}' of tree $tree should not exist after phase ElimErasedValueType.")
+  }
+
+  // FIXME: Is using a TypeMap here the best solution?
+  def elimErasedValueType(tpe: Type)(implicit ctx: Context) = new TypeMap() {
+    def apply(tp: Type) = tp match {
+      case ErasedValueType(_, underlying) =>
+        underlying
+      case _ =>
+        mapOver(tp)
+    }
+  }.apply(tpe)
+
+  def transformTypeOfTree(tree: Tree)(implicit ctx: Context): Tree =
+    tree.withType(elimErasedValueType(tree.tpe))
+
+  override def transformApply(tree: Apply)(implicit ctx: Context, info: TransformerInfo): Tree = {
+    val Apply(fun, args) = tree
+    val name = fun.symbol.name
+
+    // The casts to and from ErasedValueType are no longer needed once ErasedValueType
+    // has been eliminated.
+    val t =
+      if ((name eq nme.UNDERLYING2EVT) || (name eq nme.EVT2UNDERLYING))
+        args.head
+      else
+        tree
+    transformTypeOfTree(t)
+  }
+
+  // FIXME: This is probably missing some nodes, all the current ones were added to make tests pass
+  override def transformIdent(tree: Ident)(implicit ctx: Context, info: TransformerInfo): Tree =
+    transformTypeOfTree(tree)
+  override def transformSelect(tree: Select)(implicit ctx: Context, info: TransformerInfo): Tree =
+    transformTypeOfTree(tree)
+  override def transformBlock(tree: Block)(implicit ctx: Context, info: TransformerInfo): Tree =
+    transformTypeOfTree(tree)
```

Comment:
If I just disable `transformBlock` and `transformIf`, then `dotc_core`, `dotc_util` and `dotc_core_pickling` fail whereas they didn't fail before. Here is a minimal example which fails without `transformIf`:
```scala
class Position(val foo: Int) extends AnyVal {
  def orElse(that: Position) =
    if (this != null) this else that
}
```
Backtrace:
```scala
run -Ycheck:elimerasedvaluetype try/vcif.scala
[info] Running dotty.tools.dotc.Main -Ycheck:elimerasedvaluetype try/vcif.scala
[info] checking try/vcif.scala after phase TreeTransform:{elimerasedvaluetype, peepholeoptimize, mixin, memoize, capturedVars, constructors}
[info] exception while typing final def orElse$extension($this: Int, that: Int): Int = 
[info]   if new Position($this).equals(null).unary_!() then $this else that of class class dotty.tools.dotc.ast.Trees$DefDef # 548
[info] exception while typing final module class Position$ extends Object { this: Position$ => 
[info]   def <init>(): Position$ = {
[info]     super()
[info]     ()
[info]   }
[info]   final def orElse$extension($this: Int, that: Int): Int = 
[info]     if new Position($this).equals(null).unary_!() then $this else that
[info]   override final def hashCode$extension($this: Int): Int = 
[info]     scala.Int.box($this).hashCode()
[info]   override final def equals$extension($this: Int, x$0: Object): Boolean = {
[info]     case val selector11: Object = x$0
[info]     {
[info]       def case11(): Boolean = {
[info]         def case21(): Boolean = {
[info]           def matchFail11(): Boolean = throw new MatchError(selector11)
[info]           if true then {
[info]             false
[info]           } else matchFail11()
[info]         }
[info]         if selector11.isInstanceOf[Position] then {
[info]           val x$0: Int = {
[info]             val ev$1: Object = selector11
[info]             if ev$1.eq(null) then scala.Int.unbox(null) else 
[info]               ev$1.asInstanceOf[Position].foo()
[info]           }
[info]           {
[info]             scala.Int.box($this).equals(scala.Int.box(x$0))
[info]           }
[info]         } else case21()
[info]       }
[info]       case11()
[info]     }
[info]   }
[info] } of class class dotty.tools.dotc.ast.Trees$TypeDef # 588
[info] exception while typing package <empty> {
[info]   class Position extends Object { 
[info]     def <init>(foo: Int): Position = {
[info]       super()
[info]       this.foo$$local = foo
[info]       ()
[info]     }
[info]     private val foo$$local: Int
[info]     <accessor> def foo(): Int = this.foo$$local
[info]     def orElse(that: Int): Int = 
[info]       Position.orElse$extension(Position.this.foo(), that)
[info]     override def hashCode(): Int = 
[info]       Position.hashCode$extension(Position.this.foo())
[info]     override def equals(x$0: Object): Boolean = 
[info]       Position.equals$extension(Position.this.foo(), x$0)
[info]   }
[info]   final lazy module val Position: Position$ = new Position$()
[info]   final module class Position$ extends Object { this: Position$ => 
[info]     def <init>(): Position$ = {
[info]       super()
[info]       ()
[info]     }
[info]     final def orElse$extension($this: Int, that: Int): Int = 
[info]       if new Position($this).equals(null).unary_!() then $this else that
[info]     override final def hashCode$extension($this: Int): Int = 
[info]       scala.Int.box($this).hashCode()
[info]     override final def equals$extension($this: Int, x$0: Object): Boolean = {
[info]       case val selector11: Object = x$0
[info]       {
[info]         def case11(): Boolean = {
[info]           def case21(): Boolean = {
[info]             def matchFail11(): Boolean = throw new MatchError(selector11)
[info]             if true then {
[info]               false
[info]             } else matchFail11()
[info]           }
[info]           if selector11.isInstanceOf[Position] then {
[info]             val x$0: Int = {
[info]               val ev$1: Object = selector11
[info]               if ev$1.eq(null) then scala.Int.unbox(null) else 
[info]                 ev$1.asInstanceOf[Position].foo()
[info]             }
[info]             {
[info]               scala.Int.box($this).equals(scala.Int.box(x$0))
[info]             }
[info]           } else case21()
[info]         }
[info]         case11()
[info]       }
[info]     }
[info]   }
[info] } of class class dotty.tools.dotc.ast.Trees$PackageDef # 589
[info] *** error while checking after phase constructors ***
[error] Exception in thread "main" java.lang.AssertionError: assertion failed: Types differ
[error] Original type : ErasedValueType(Position, Int)
[error] After checking: Int
[error] Original tree : if new Position($this).equals(null).unary_!() then $this else that
[error] After checking: if new Position($this).equals(null).unary_!() then $this else that
[error] Why different :
[error]              Subtype trace:
[error]   ==> Int <:< ErasedValueType(Position, Int) class dotty.tools.dotc.core.Types$CachedTypeRef class dotty.tools.dotc.core.TypeErasure$CachedErasedValueType
[error]   <== Int <:< ErasedValueType(Position, Int) class dotty.tools.dotc.core.Types$CachedTypeRef class dotty.tools.dotc.core.TypeErasure$CachedErasedValueType = false
[error]         at scala.Predef$.assert(Predef.scala:165)
[error]         at dotty.tools.dotc.transform.TreeChecker$Checker.typedUnadapted(TreeChecker.scala:179)
[error]         at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1088)
[error]         at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1086)
[error]         at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:147)
[error]         at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:51)
[error]         at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1086)
[error]         at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1125)
[error]         at dotty.tools.dotc.typer.Typer$$anonfun$typedDefDef$1.apply(Typer.scala:892)
[error]         at dotty.tools.dotc.typer.Typer$$anonfun$typedDefDef$1.apply(Typer.scala:885)
[error]         at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
[error]         at dotty.tools.dotc.typer.Typer.typedDefDef(Typer.scala:885)
[error]         at dotty.tools.dotc.transform.TreeChecker$Checker.dotty$tools$dotc$transform$TreeChecker$Checker$$super$typedDefDef(TreeChecker.scala:249)
[error]         at dotty.tools.dotc.transform.TreeChecker$Checker$$anonfun$typedDefDef$1$$anonfun$apply$3.apply(TreeChecker.scala:249)
[error]         at dotty.tools.dotc.transform.TreeChecker$Checker$$anonfun$typedDefDef$1$$anonfun$apply$3.apply(TreeChecker.scala:249)
[error]         at dotty.tools.dotc.core.Decorators$ListDecorator$.foldRightBN$extension(Decorators.scala:104)
[error]         at dotty.tools.dotc.core.Decorators$ListDecorator$$anonfun$foldRightBN$extension$1.apply(Decorators.scala:105)
[error]         at dotty.tools.dotc.core.Decorators$ListDecorator$.foldRightBN$extension(Decorators.scala:104)
[error]         at dotty.tools.dotc.core.Decorators$ListDecorator$$anonfun$foldRightBN$extension$1.apply(Decorators.scala:105)
[error]         at dotty.tools.dotc.transform.TreeChecker$Checker.withDefinedSym(TreeChecker.scala:139)
[error]         at dotty.tools.dotc.transform.TreeChecker$Checker$$anonfun$withDefinedSyms$1.apply(TreeChecker.scala:147)
[error]         at dotty.tools.dotc.transform.TreeChecker$Checker$$anonfun$withDefinedSyms$1.apply(TreeChecker.scala:147)
[error]         at dotty.tools.dotc.core.Decorators$ListDecorator$.foldRightBN$extension(Decorators.scala:105)
[error]         at dotty.tools.dotc.core.Decorators$ListDecorator$$anonfun$foldRightBN$extension$1.apply(Decorators.scala:105)
[error]         at dotty.tools.dotc.transform.TreeChecker$Checker.withDefinedSym(TreeChecker.scala:139)
[error]         at dotty.tools.dotc.transform.TreeChecker$Checker$$anonfun$withDefinedSyms$1.apply(TreeChecker.scala:147)
[error]         at dotty.tools.dotc.transform.TreeChecker$Checker$$anonfun$withDefinedSyms$1.apply(TreeChecker.scala:147)
[error]         at dotty.tools.dotc.core.Decorators$ListDecorator$.foldRightBN$extension(Decorators.scala:105)
[error]         at dotty.tools.dotc.transform.TreeChecker$Checker.withDefinedSyms(TreeChecker.scala:147)
[error]         at dotty.tools.dotc.transform.TreeChecker$Checker$$anonfun$withDefinedSymss$1.apply(TreeChecker.scala:150)
[error]         at dotty.tools.dotc.transform.TreeChecker$Checker$$anonfun$withDefinedSymss$1.apply(TreeChecker.scala:150)
[error]         at dotty.tools.dotc.core.Decorators$ListDecorator$.foldRightBN$extension(Decorators.scala:105)
[error]         at dotty.tools.dotc.transform.TreeChecker$Checker.withDefinedSymss(TreeChecker.scala:150)
[error]         at dotty.tools.dotc.transform.TreeChecker$Checker$$anonfun$typedDefDef$1.apply(TreeChecker.scala:248)
[error]         at dotty.tools.dotc.transform.TreeChecker$Checker$$anonfun$typedDefDef$1.apply(TreeChecker.scala:248)
[error]         at dotty.tools.dotc.core.Decorators$ListDecorator$.foldRightBN$extension(Decorators.scala:104)
[error]         at dotty.tools.dotc.transform.TreeChecker$Checker.withDefinedSyms(TreeChecker.scala:147)
[error]         at dotty.tools.dotc.transform.TreeChecker$Checker.typedDefDef(TreeChecker.scala:247)
[error]         at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1026)
[error]         at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1076)
[error]         at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:93)
[error]         at dotty.tools.dotc.transform.TreeChecker$Checker.typedUnadapted(TreeChecker.scala:166)
[error]         at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1088)
[error]         at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1086)
[error]         at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:147)
[error]         at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:51)
[error]         at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1086)
[error]         at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:1110)
[error]         at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1121)
[error]         at dotty.tools.dotc.transform.TreeChecker$Checker.typedStats(TreeChecker.scala:275)
[error]         at dotty.tools.dotc.typer.Typer$$anonfun$typedClassDef$1.apply(Typer.scala:922)
[error]         at dotty.tools.dotc.typer.Typer$$anonfun$typedClassDef$1.apply(Typer.scala:904)
[error]         at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
[error]         at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:904)
[error]         at dotty.tools.dotc.transform.TreeChecker$Checker.typedClassDef(TreeChecker.scala:243)
[error]         at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1029)
[error]         at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1076)
[error]         at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:93)
[error]         at dotty.tools.dotc.transform.TreeChecker$Checker.typedUnadapted(TreeChecker.scala:166)
[error]         at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1088)
[error]         at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1086)
[error]         at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:147)
[error]         at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:51)
[error]         at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1086)
[error]         at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:1110)
[error]         at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1121)
[error]         at dotty.tools.dotc.transform.TreeChecker$Checker.typedStats(TreeChecker.scala:275)
[error]         at dotty.tools.dotc.typer.Typer$$anonfun$typedPackageDef$1.apply(Typer.scala:969)
[error]         at dotty.tools.dotc.typer.Typer$$anonfun$typedPackageDef$1.apply(Typer.scala:960)
[error]         at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
[error]         at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:960)
[error]         at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1066)
[error]         at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1078)
[error]         at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:93)
[error]         at dotty.tools.dotc.transform.TreeChecker$Checker.typedUnadapted(TreeChecker.scala:166)
[error]         at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1088)
[error]         at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1086)
[error]         at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:147)
[error]         at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:51)
[error]         at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1086)
[error]         at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1125)
[error]         at dotty.tools.dotc.transform.TreeChecker.check(TreeChecker.scala:100)
[error]         at dotty.tools.dotc.transform.TreeChecker.run(TreeChecker.scala:78)
[error]         at dotty.tools.dotc.core.Phases$Phase$$anonfun$runOn$1.apply(Phases.scala:265)
[error]         at dotty.tools.dotc.core.Phases$Phase$$anonfun$runOn$1.apply(Phases.scala:263)
[error]         at scala.collection.immutable.List.map(List.scala:273)
[error]         at dotty.tools.dotc.core.Phases$Phase$class.runOn(Phases.scala:263)
[error]         at dotty.tools.dotc.transform.TreeChecker.runOn(TreeChecker.scala:38)
[error]         at dotty.tools.dotc.Run$$anonfun$compileSources$1$$anonfun$apply$mcV$sp$3.apply(Run.scala:50)
[error]         at dotty.tools.dotc.Run$$anonfun$compileSources$1$$anonfun$apply$mcV$sp$3.apply(Run.scala:47)
[error]         at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
[error]         at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
[error]         at dotty.tools.dotc.Run$$anonfun$compileSources$1.apply$mcV$sp(Run.scala:47)
[error]         at dotty.tools.dotc.Run$$anonfun$compileSources$1.apply(Run.scala:42)
[error]         at dotty.tools.dotc.Run$$anonfun$compileSources$1.apply(Run.scala:42)
[error]         at dotty.tools.dotc.util.Stats$.monitorHeartBeat(Stats.scala:68)
[error]         at dotty.tools.dotc.Run.compileSources(Run.scala:41)
[error]         at dotty.tools.dotc.Run.compile(Run.scala:32)
[error]         at dotty.tools.dotc.Driver.doCompile(Driver.scala:20)
[error]         at dotty.tools.dotc.Main$.doCompile(Main.scala:26)
[error]         at dotty.tools.dotc.Driver.process(Driver.scala:31)
[error]         at dotty.tools.dotc.Driver.main(Driver.scala:40)
[error]         at dotty.tools.dotc.Main.main(Main.scala)
```